docs/radio-bridge-protocol.md

```diff
@@ -0,0 +1,93 @@
+# Radio Bridge Protocol
+
+## Context
+
+The current method of connecting a micro:bit to the ML web app is via Web Bluetooth. The issue with this is that Web Bluetooth may not be enabled/available on many systems that are used in schools, which is the primary use-case for the ML trainer web app. It would also be beneficial to be able to handle multiple micro:bit connections without having to change the design of the web app too much from its current state.
+
+## Proposed solution
+
+The solution to this could be a single "bridge" Micro:bit connected to a host computer via WebUSB. Micro:bits used in the classroom that need to connect to the ML trainer web app would then connect to the bridge Micro:bit via the Micro:bits' radio connection. The bridge would recieve streaming data from one Micro:bit in the classroom at a time and forward this data onto the web app using a message format that we have designed. The connections to the Micro:bits in the classroom are managed solely by the bridge Micro:bit.
+
+## The message format
+
+Data is streamed to/from the Micro:bit and web app via serial data over USB.
+
+### Delimiter
+
+The messages will need to be delimited by a character for example a newline.
+
+```
+\n
+```
+
+### Handshake request from web app, response from Micro:bit
+
+In order to connect the bridge Micro:bit to the web app, a handshake should be performed. This confirms that the bridge Micro:bit is already running the correct binary, so the flashing step for the Micro:bit can be skipped.
+
+Handshake request:
+```
+Req
+```
+
+Handshake response
+```
+Ack
+```
+
+### Sending LED matrix data
+
+The web app has "triggers" that send LED data to the Micro:bit when a particular action is detected.
+
+```
+LED[1,1,0,1,1,0,0,1,1,1,1,0,1,1,1,0,1,1,1,0,1,1,0,1,1]
```

Comment:
Do we want a flat list here, or break it down by row, e.g.

```
LED_ROW1[1,1,0,1,1,],LED_ROW2[1,1,0,1,1,],LED_ROW3[1,1,0,1,1,],LED_ROW4[1,1,0,1,1,],LED_ROW5[1,1,0,1,1,]
```
Or something else?

Reply:
Having a single entry will save characters, so that might be a slight advantage.
Although having commas inside the brackets would make parsing the string more difficult, so we either remove the commas to separate the different entries, we remove the commas inside the square brackets or we use a different character inside the square brackets.

One thing to also take into consideration is that the brightness of each LED can also be changed from 0 to 255.

And we also need to consider scrolling strings and things like that.

Reply:
Should we have something like LED_ROW1[0,255,128,245,1] then? I think if we have these values we'd need a delineator for the values? In this case ",".

Reply:
Yes, being able to have commas between the square brackets I was thinking we can do `LED[11011,11011,11011,11011,11011]`